app/src/main/java/org/dicekeys/app/fragments/dicekey/SoloKeyFragment.kt

```diff
@@ -1,21 +1,230 @@
 package org.dicekeys.app.fragments.dicekey
 
+import android.content.Intent
+import android.content.IntentFilter
+import android.hardware.usb.UsbDevice
+import android.hardware.usb.UsbManager
 import android.os.Bundle
+import android.os.CountDownTimer
 import android.view.View
+import androidx.core.view.isVisible
+import androidx.core.widget.doAfterTextChanged
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.Observer
+import com.dicekeys.fidowriter.ListOfWritableUsbFidoKeys
 import dagger.hilt.android.AndroidEntryPoint
+import org.dicekeys.api.DerivationRecipe
+import org.dicekeys.api.derivationRecipeTemplates
 import org.dicekeys.app.AppFragment
 import org.dicekeys.app.R
 import org.dicekeys.app.databinding.SolokeyFragmentBinding
 import org.dicekeys.app.viewmodels.DiceKeyViewModel
+import org.dicekeys.crypto.seeded.Secret
 
 @AndroidEntryPoint
 class SoloKeyFragment: AppFragment<SolokeyFragmentBinding>(R.layout.solokey_fragment) {
 
+    private lateinit var deviceList: ListOfWritableUsbFidoKeys
+    private lateinit var permissionIntent: android.app.PendingIntent
     private lateinit var viewModel: DiceKeyViewModel
+    private val INTENT_ACTION_USB_PERMISSION_EVENT = "org.dicekeys.intents.USB_PERMISSION_EVENT"
+    private var isWriteUnderway: Boolean = false
+    private var derivationRecipe = MutableLiveData<DerivationRecipe>()
+    private var sequenceNumber = MutableLiveData<Int>(1)
 
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
         super.onViewCreated(view, savedInstanceState)
 
         viewModel = getDiceKeyRootFragment().viewModel
+
+        binding.vm = viewModel
+
+        /*Get selected derivationRecipe template*/
+        derivationRecipe.value = derivationRecipeTemplates.get(1)
+
+        binding.tvRecipeJson.text = derivationRecipe.value!!.derivationOptionsJson
```

Comment:
Prefer the use of a viewModel here and connect the data values in xml.
You can have multiple viewmodels, one from the parent fragment and one for solokey data.

It will greatly reduce your code.